### Task 3:
***
November 16th, 2020: The standard deviation of an array of numbers x is
calculated using numpy as np.sqrt(np.sum((x - np.mean(x))**2)/len(x)) .
However, Microsoft Excel has two different versions of the standard deviation calculation, STDDEV.P and STDEV.S . The STDDEV.P function performs the above
calculation but in the STDEV.S calculation the division is by len(x)-1 rather
than len(x) . Research these Excel functions, writing a note in a Markdown cell
about the difference between them. Then use numpy to perform a simulation
demonstrating that the STDEV.S calculation is a better estimate for the standard deviation of a population when performed on a sample. Note that part of
this task is to figure out the terminology in the previous sentence.
***

Standard deviation is a measure of variability and measures the average distance a number is from its mean.

The STDEV.P function is used to calculate the standard deviation of the whole population.[3.1] The P stands for population.  This will deal with the accuracy issue of STDEV.S.

$$ \sigma  = \sqrt \frac{\sum {\left( {x_i  - \mu} \right)^2 } } {N} $$ [3.2]

To calculate the standard deviation Sigma (population standard deviation) equals the square root of the sum of squares using  mu (the population mean), divided by the sample size.


The Excel STDEV.S Function is used to calculate the standard deviation in the sample data set. The S in STDEV.S stands for Sample.  It means there is more data out there than we currently have. So the real standard deviation of the data is not known. We have a piece of it so will try and calculate it from the piece we do have. [3.3]  A sample may be used as it is easier and cheaper than having to to look at all of the population but does not take into account the possible bias if the sample set is taken which doesnt adequately represent the full population so accuracy is lost. And maybe there is no access to the entire population.

The formula for Sample Standard Deviation is:
$$ s  = \sqrt \frac{\sum {\left( {x_i  - \bar x} \right)^2 } } {n-1} $$ [3.2]

S (sample standard deviation) is equal to the square root of the sum of squares using the sample mean x-bar divided by the sample size minus one.

If the same values were given for both formulas the sample standard deviation will result in a higher value for the standard deviation as the sample size minus one is the denominator.  A standard deviation of a sample can underestimate the true population standard deviation so when the formula is used, estimating the population standard deviation using a sample, the standard deviation for the sample itself is still the population standard deviation.  The true standard deviation for the results would use the STDEV.P formula.  The STDEV.S would be used when you want to estimate what the true population standard deviation is based on a sample.




https://honingds.com/blog/python-standard-deviation/


numpy uses population standard deviation by default, which is similar to pstdev of statistics module.

If you want to use it to calculate sample standard deviation, use an additional parameter, called ddof and set it to 1.

By default ddof is 0.

Let’s now calculate standard deviation using numpy.std for a sample of the data.

This result will be same as what we got when we used stdev of statistics module.

***
##### Taking a look at the various ways to calculate STDEV.P and STDEV.S and see if the results are the same.


In [1]:
import numpy as np
# Set the range of numbers to be used in different formulas and call it x
x =  ([9,2,5,4,12,7,8,11,9,3,7,4,12,5,4,10,9,6,9,4])
# Returns the standard deviation (same as STDEV.P)
np.std (x)

2.9832867780352594

In [2]:
# Calculate sample standard deviation (STDEV.S), use an additional parameter, called ddof and set it to 1. [3.4]
np.std(x, ddof = 1)

3.0607876523260447

In [3]:
import statistics
# Calculate the standard deviation, (STDEV.S). stdev() function only calculates standard deviation from a sample of data, rather than an entire population. [3.5]
statistics.stdev(x)


3.0607876523260447

In [4]:
# Calculate standard deviation of an entire population (STDEV.P)
statistics.pstdev(x)

2.9832867780352594

In [5]:
#  Use formula given by lecturer on assessment sheet to calculate STDEV.P 
np.sqrt(np.sum((x - np.mean(x))**2)/len(x))

2.9832867780352594

In [6]:
# Since I know that the below calculates STDEV.S, use it as a double check future results
np.std (x, ddof = 1)

3.0607876523260447

In [7]:
# Amend forumla given by lecturer to calculate STDEV.S and check result against above.
np.sqrt(np.sum((x - np.mean(x))**2)/(len(x)-1))


3.0607876523260447

***
#### Using numpy to perform a simulation demonstrating that the STDEV.S calculation is a better estimate for the standard deviation of a population when performed on a sample.

In [8]:
# Calculate Standard Deviation of entire population and call it x2
x2=([54, 45, 59, 51, 52, 49, 41, 50, 53, 45, 36, 59, 51, 47, 55, 37, 51, 58, 57, 52, 40, 44, 42, 47, 51, 53, 30, 36, 34, 49, 51, 49, 40, 52, 55, 41, 44, 50, 61, 34])
np.sqrt(np.sum((x2 - np.mean(x2))**2)/len(x2))

7.604891517963948

Sample Deviation of set x2, above, taking first 15 numbers.
Assume only had access to part of the population a sample of the first 15 numbers.

The entire population (x2 above) has a standard deviation of 7.6048. 

Taking the first 15 records of x2 dataset and using the sample standard deviation formula, the sample standard deviation will generate a higher value for the standard deviation due to the sample size minus one being the denominator.  STDEV.S will underestimate the true population standard deviation.  

Applying the formula will estimate the population standard deviation using a sample.  The population standard deviation is the true standard deviation for the values.  The sample standard deviation is used to estimate the true population standard deviation based on a sample.

In [12]:
# Create sample of above set x2 (first 15 numbers) and call it x3
x3=([54, 45, 59, 51, 52, 49, 41, 50, 53, 45, 36, 59, 51, 47, 55])
# Population standard deviation (STDEV.P) of the sample is
np.sqrt(np.sum((x3 - np.mean(x3))**2)/len(x3))

6.079473661428267

In [10]:
# Sample standard deviation (STDEV.S) of the same sample set above (x3)
np.sqrt(np.sum((x3 - np.mean(x3))**2)/(len(x3)-1))

6.292853089020911

#### The result is that the STDEV.S (6.292853089020911) is higher than the STDEV.P (6.079473661428267) of the sample, but STDEV.S is closer to the actual population standard deviation of 7.604891517963948.  


[3.1] Exceltip, exceltip.com, How To Use Excel STDEV.P Function, https://www.exceltip.com/statistical-formulas/how-to-use-excel-stdev-p-function.html#:~:text=Let's%20Explore.-,The%20STDEV.,sample%20of%20the%20entire%20population., accessed November 2020 <br>

[3.2] Excel Easy, excel-easy.com, Standard Deviation, https://www.excel-easy.com/examples/standard-deviation.html, accessed November 2020 <br>

[3.3] Excetip, exceltip.com, How to use the STDEV.S Function in Excel, https://www.exceltip.com/statistical-formulas/how-to-use-stdev-s-function-in-excel.html, accessed November 2020 <br>

[3.4] Mad Physicist, (2015), stackoverflow.com, Standard deviation in numpy, https://stackoverflow.com/questions/34050491/standard-deviation-in-numpy, accessed November 2020 <br>

[3.5] retr0, (2020), geeksforgeeks.org, stdev() method in Python statistics module, https://www.geeksforgeeks.org/python-statistics-stdev/, accessed November 2020 <br>

Further reading:

* Dr. Todd Grande
, 2016, youtube.com, Understanding Sample (STDEV.S) and Population (STDEV.P) Standard Deviation using Excel, https://www.youtube.com/watch?v=W7q8kfs1bNI&ab_channel=Dr.ToddGrande, accessed November 2020 <br>

* (2017), mathisisfun.com, Standard Deviation Formulas, https://www.mathsisfun.com/data/standard-deviation-formulas.html, accessed November 2020 <br>

* exceptip.com, How To Use Excel STDEV.P Function, https://www.exceltip.com/statistical-formulas/how-to-use-excel-stdev-p-function.html#:~:text=Let's%20Explore.-,The%20STDEV.,sample%20of%20the%20entire%20population., accessed November 2020 <br>
